File: examples/js/src/arithmetic.js
// The six algorithms the seal rests on, named once and never negotiated:
// Ed25519 signs, X25519 agrees, ML-KEM-768 encapsulates, SHA-256 hashes,
// AES-256-GCM encrypts and HKDF-SHA-256 derives. A secret is always the
// thirty-two bytes a key is made from, and a lock the sixty-four, d then z.

import { ml_kem768 } from '@noble/post-quantum/ml-kem.js';

const subtle = globalThis.crypto.subtle;

export const hex = (bytes) => Array.from(bytes, (b) => b.toString(16).padStart(2, '0')).join('');
export const unhex = (s) => Uint8Array.from(s.match(/../g) ?? [], (h) => parseInt(h, 16));
export const utf8 = (text) => new TextEncoder().encode(text);

export function concat(...parts) {
  const out = new Uint8Array(parts.reduce((n, p) => n + p.length, 0));
  let at = 0;
  for (const p of parts) {
    out.set(p, at);
    at += p.length;
  }
  return out;
}

export async function sha256(bytes) {
  return new Uint8Array(await subtle.digest('SHA-256', bytes));
}

// HKDF-SHA-256 under the zero-length salt, with the label's ASCII bytes as info.
export async function hkdf(ikm, label, length) {
  const key = await subtle.importKey('raw', ikm, 'HKDF', false, ['deriveBits']);
  const info = typeof label === 'string' ? utf8(label) : label;
  const bits = await subtle.deriveBits({ name: 'HKDF', hash: 'SHA-256', salt: new Uint8Array(0), info }, key, length * 8);
  return new Uint8Array(bits);
}

// WebCrypto takes a raw private key only wrapped in PKCS#8; these are the
// fixed prefixes for a 32-byte Ed25519 seed and a 32-byte X25519 scalar.
const PKCS8 = {
  Ed25519: unhex('302e020100300506032b657004220420'),
  X25519: unhex('302e020100300506032b656e04220420'),
};

const privateKey = (name, secret, usages) => subtle.importKey('pkcs8', concat(PKCS8[name], secret), { name }, true, usages);

async function publicOf(name, secret, usages) {
  const { x } = await subtle.exportKey('jwk', await privateKey(name, secret, usages));
  return Uint8Array.from(atob(x.replace(/-/g, '+').replace(/_/g, '/')), (c) => c.charCodeAt(0));
}

// Ed25519 signs. Verification is RFC 8032's cofactorless check, and it refuses
// in exactly four places: a signature not sixty-four bytes, an s at or above
// the group order, an R that is not how its point encodes, and a public key
// that is small-order or whose y is not reduced.
const P = 2n ** 255n - 19n;
const L = 2n ** 252n + 27742317777372353535851937790883648493n;
const littleEndian = (bytes) => bytes.reduceRight((n, b) => (n << 8n) | BigInt(b), 0n);
const yOf = (point) => littleEndian(point) & ((1n << 255n) - 1n);
const SMALL_ORDER_Y = new Set([
  0n,
  1n,
  P - 1n,
  yOf(unhex('26e8958fc2b227b045c3f489f2ef98f0d5dfac05d3c63339b13802886d53fc05')),
  yOf(unhex('c7176a703d4dd84fba3c0b760d10670f2a2053fa2c39ccc64ec7fd7792ac037a')),
]);

export const ed25519 = {
  publicKey: (secret) => publicOf('Ed25519', secret, ['sign']),

  async sign(secret, message) {
    return new Uint8Array(await subtle.sign('Ed25519', await privateKey('Ed25519', secret, ['sign']), message));
  },

  async verify(publicKey, message, signature) {
    if (publicKey.length !== 32 || signature.length !== 64) return false;
    const y = yOf(publicKey);
    if (y >= P || SMALL_ORDER_Y.has(y)) return false;
    if (yOf(signature.subarray(0, 32)) >= P) return false;
    if (littleEndian(signature.subarray(32)) >= L) return false;
    // The equation itself, `[s]B = R + [k]A`, computed here: a library's own
    // check may refuse where RFC 8032's cofactorless check does not, a
    // small-order R among them. [s]B - [k]A is encoded and compared with R
    // as it was received.
    const A = decodePoint(publicKey);
    if (!A) return false;
    const R = signature.subarray(0, 32);
    const k = littleEndian(new Uint8Array(await subtle.digest('SHA-512', concat(R, publicKey, message)))) % L;
    const s = littleEndian(signature.subarray(32));
    const check = encodePoint(addPoints(multiply(s, BASE), negate(multiply(k, A))));
    return check.every((b, i) => b === R[i]);
  },
};

// Edwards25519 in extended coordinates, RFC 8032 section 5.1.
const mod = (x) => ((x % P) + P) % P;
function power(b, e) {
  let r = 1n;
  b = mod(b);
  while (e > 0n) {
    if (e & 1n) r = (r * b) % P;
    b = (b * b) % P;
    e >>= 1n;
  }
  return r;
}
const inverse = (x) => power(x, P - 2n);
const D = mod(-121665n * inverse(121666n));
const SQRT_M1 = power(2n, (P - 1n) / 4n);

function decodePoint(bytes) {
  const y = yOf(bytes);
  if (y >= P) return null;
  const sign = bytes[31] >> 7;
  const x2 = mod((y * y - 1n) * inverse(D * y * y + 1n));
  let x = power(x2, (P + 3n) / 8n);
  if (mod(x * x - x2) !== 0n) x = mod(x * SQRT_M1);
  if (mod(x * x - x2) !== 0n) return null;
  if (x === 0n && sign === 1) return null;
  if (Number(x & 1n) !== sign) x = P - x;
  return [x, y, 1n, mod(x * y)];
}

function encodePoint([X, Y, Z]) {
  const zi = inverse(Z);
  const x = mod(X * zi);
  let y = mod(Y * zi);
  const out = new Uint8Array(32);
  for (let i = 0; i < 32; i++) {
    out[i] = Number(y & 0xffn);
    y >>= 8n;
  }
  out[31] |= Number(x & 1n) << 7;
  return out;
}

function addPoints([X1, Y1, Z1, T1], [X2, Y2, Z2, T2]) {
  const a = mod((Y1 - X1) * (Y2 - X2));
  const b = mod((Y1 + X1) * (Y2 + X2));
  const c = mod(T1 * 2n * D * T2);
  const d = mod(Z1 * 2n * Z2);
  const [e, f, g, h] = [b - a, d - c, d + c, b + a];
  return [mod(e * f), mod(g * h), mod(f * g), mod(e * h)];
}

const negate = ([X, Y, Z, T]) => [mod(-X), Y, Z, mod(-T)];

function multiply(n, point) {
  let result = [0n, 1n, 1n, 0n];
  let q = point;
  while (n > 0n) {
    if (n & 1n) result = addPoints(result, q);
    q = addPoints(q, q);
    n >>= 1n;
  }
  return result;
}

const BASE = decodePoint(unhex('5866666666666666666666666666666666666666666666666666666666666666'));

// X25519 agrees. An agreement that comes out all zero is no agreement at all.
export const x25519 = {
  publicKey: (secret) => publicOf('X25519', secret, ['deriveBits']),

  async agree(secret, publicKey) {
    try {
      const pub = await subtle.importKey('raw', publicKey, 'X25519', false, []);
      const priv = await privateKey('X25519', secret, ['deriveBits']);
      const shared = new Uint8Array(await subtle.deriveBits({ name: 'X25519', public: pub }, priv, 256));
      return shared.some((b) => b !== 0) ? shared : null;
    } catch {
      return null;
    }
  },
};

// ML-KEM-768 in its deterministic internal forms, FIPS 203: a key pair from d
// and z, a ciphertext and shared secret from an encapsulation key and m, and
// decapsulation with implicit rejection. A ciphertext of another length is no
// secret at all.
export const mlkem = {
  keygen(dz) {
    const { publicKey, secretKey } = ml_kem768.keygen(dz);
    return { ek: publicKey, dk: secretKey };
  },

  encaps(ek, m) {
    const { cipherText, sharedSecret } = ml_kem768.encapsulate(ek, m);
    return { ciphertext: cipherText, shared: sharedSecret };
  },

  decaps(dk, ciphertext) {
    try {
      return ml_kem768.decapsulate(ciphertext, dk);
    } catch {
      return null;
    }
  },
};

// The message cipher: forty-four bytes from HKDF, the first thirty-two the
// AES-256 key and the last twelve the nonce. The label is `quo-seal` over an
// agreement, and `quo-edge-seal` over an agreement then an edge key for the
// body of an ask. The nonce needs no randomness of its own, because the key
// beside it is fresh on every message.
async function messageKey(ikm, label) {
  const out = await hkdf(ikm, label, 44);
  const key = await subtle.importKey('raw', out.subarray(0, 32), 'AES-GCM', false, ['encrypt', 'decrypt']);
  return { key, iv: out.subarray(32) };
}

export async function encrypt(ikm, additionalData, plaintext, label = 'quo-seal') {
  const { key, iv } = await messageKey(ikm, label);
  return new Uint8Array(await subtle.encrypt({ name: 'AES-GCM', iv, additionalData }, key, plaintext));
}

export async function decrypt(ikm, additionalData, ciphertext, label = 'quo-seal') {
  const { key, iv } = await messageKey(ikm, label);
  try {
    return new Uint8Array(await subtle.decrypt({ name: 'AES-GCM', iv, additionalData }, key, ciphertext));
  } catch {
    return null;
  }
}

File: examples/rust/src/arithmetic.rs
// SPDX-License-Identifier: Apache-2.0
//! The five algorithms and the derivations the spec names over them.
//!
//! Ed25519 signs, X25519 agrees, ML-KEM-768 encapsulates, SHA-256 hashes,
//! AES-256-GCM encrypts, and HKDF-SHA-256 under an empty salt derives: the
//! ward's two secrets from its seed, under `quo-ward-sign` and
//! `quo-ward-seal`, the message cipher's key and nonce from an agreement,
//! under `quo-seal`, an ask body's from an agreement and an edge key, under
//! `quo-edge-seal`, the knock's edge key from an ML-KEM secret, under
//! `quo-lock`, and the next edge key from the edge key an ask came under and
//! its reply's agreement, under `quo-edge`.

use aes_gcm::aead::{AeadInPlace, KeyInit, Nonce};
use aes_gcm::Aes256Gcm;
use ed25519_dalek::{Signature, Signer, SigningKey, Verifier, VerifyingKey};
use hkdf::SimpleHkdf;
use ml_kem::kem::Decapsulate;
use ml_kem::{Ciphertext, EncapsulateDeterministic, Encoded, EncodedSizeUser, KemCore, MlKem768, B32};
use sha2::{Digest, Sha256};

/// The label of the message cipher's derivation.
pub const LABEL_SEAL: &[u8] = b"quo-seal";
/// The label of an ask body's cipher, over the agreement and the edge key.
pub const LABEL_EDGE_SEAL: &[u8] = b"quo-edge-seal";
/// The label the next edge key is derived under.
pub const LABEL_EDGE: &[u8] = b"quo-edge";
/// The label the knock's edge key is derived under, from the ML-KEM secret.
pub const LABEL_LOCK: &[u8] = b"quo-lock";
/// The label the ward's Ed25519 secret is derived under.
pub const LABEL_WARD_SIGN: &[u8] = b"quo-ward-sign";
/// The label the ward's X25519 secret, its padlock, is derived under.
pub const LABEL_WARD_SEAL: &[u8] = b"quo-ward-seal";

/// Bytes of a key, a secret, a lid, an agreement.
pub const KEY_LEN: usize = 32;
/// Bytes of a signature.
pub const SIGNATURE_LEN: usize = 64;
/// Bytes of the AES-GCM tag, at the end of the ciphertext.
pub const TAG_LEN: usize = 16;
/// Bytes of an ML-KEM-768 encapsulation key, the lock an invitation carries.
pub const LOCK_LEN: usize = 1184;
/// Bytes of an ML-KEM-768 ciphertext, which a knock box carries.
pub const CIPHERTEXT_LEN: usize = 1088;
/// Bytes drawn for a lock: d then z.
pub const LOCK_SEED_LEN: usize = 64;

/// SHA-256 of the bytes.
pub fn sha256(bytes: &[u8]) -> [u8; 32] {
    Sha256::digest(bytes).into()
}

/// HKDF-SHA-256, extract then expand, with `info` as the ASCII label and no
/// prefix. An empty `salt` is the zero-length salt of RFC 5869.
pub fn hkdf_sha256(salt: &[u8], ikm: &[u8], info: &[u8], out: &mut [u8]) {
    let (_, hkdf) = SimpleHkdf::<Sha256>::extract(Some(salt), ikm);
    hkdf.expand(info, out).expect("HKDF-SHA-256 length is within bound");
}

/// The ward key: an Ed25519 pair that signs replies and an X25519 padlock every
/// ask is sealed to, each derived from the seed under its own label.
#[derive(Clone)]
pub struct WardKey {
    sign_secret: [u8; 32],
    seal_secret: [u8; 32],
    sign_pk: [u8; 32],
    padlock: [u8; 32],
}

impl WardKey {
    /// The ward key of a thirty-two byte seed.
    pub fn from_seed(seed: &[u8; 32]) -> Self {
        let mut sign_secret = [0u8; 32];
        let mut seal_secret = [0u8; 32];
        hkdf_sha256(&[], seed, LABEL_WARD_SIGN, &mut sign_secret);
        hkdf_sha256(&[], seed, LABEL_WARD_SEAL, &mut seal_secret);
        WardKey { sign_pk: signing_pk(&sign_secret), padlock: agreement_pk(&seal_secret), sign_secret, seal_secret }
    }

    /// The ward pk on the wire: the signing pk then the padlock, 64 bytes.
    pub fn pk(&self) -> [u8; 64] {
        let mut pk = [0u8; 64];
        pk[..32].copy_from_slice(&self.sign_pk);
        pk[32..].copy_from_slice(&self.padlock);
        pk
    }

    pub fn signing_pk(&self) -> [u8; 32] {
        self.sign_pk
    }

    pub fn padlock(&self) -> [u8; 32] {
        self.padlock
    }

    /// Signs with the ward's Ed25519 secret.
    pub fn sign(&self, message: &[u8]) -> [u8; 64] {
        sign(&self.sign_secret, message)
    }

    /// Agrees with the ward's padlock secret. `None` is an agreement refused.
    pub fn agree(&self, their_pk: &[u8; 32]) -> Option<[u8; 32]> {
        agree(&self.seal_secret, their_pk)
    }
}

/// The Ed25519 public key of a 32-byte secret, taken as it stands.
pub fn signing_pk(secret: &[u8; 32]) -> [u8; 32] {
    SigningKey::from_bytes(secret).verifying_key().to_bytes()
}

/// An Ed25519 signature.
pub fn sign(secret: &[u8; 32], message: &[u8]) -> [u8; 64] {
    SigningKey::from_bytes(secret).sign(message).to_bytes()
}

/// The field prime 2^255 - 19, little-endian.
const P: [u8; 32] = {
    let mut p = [0xffu8; 32];
    p[0] = 0xed;
    p[31] = 0x7f;
    p
};

/// Whether the y coordinate of a compressed point is reduced, below the prime.
fn y_is_canonical(pk: &[u8; 32]) -> bool {
    let mut y = *pk;
    y[31] &= 0x7f;
    for i in (0..32).rev() {
        if y[i] != P[i] {
            return y[i] < P[i];
        }
    }
    false
}

/// Ed25519 verification against the bytes exactly as received: RFC 8032's
/// cofactorless check, `[s]B = R + [k]A`, with `R` compared as encoded. It
/// refuses in four places and no others: a signature of any length but
/// sixty-four, an `s` at or above the group order, an `R` whose bytes are not
/// the encoding of the point they name, and a public key that is small-order
/// in any spelling or whose y is at or above the prime. A torsion public key
/// and a small-order `R` verify.
pub fn verify(pk: &[u8], message: &[u8], signature: &[u8]) -> bool {
    let (Ok(pk), Ok(sig)) = (<[u8; 32]>::try_from(pk), <[u8; 64]>::try_from(signature)) else {
        return false;
    };
    if !y_is_canonical(&pk) {
        return false;
    }
    let Ok(key) = VerifyingKey::from_bytes(&pk) else {
        return false;
    };
    if key.is_weak() {
        return false;
    }
    // The cofactorless check: `s` refused unreduced, the recomputed `R`
    // compared with the bytes received.
    key.verify(message, &Signature::from_bytes(&sig)).is_ok()
}

/// The X25519 public key of a 32-byte secret.
pub fn agreement_pk(secret: &[u8; 32]) -> [u8; 32] {
    let secret = x25519_dalek::StaticSecret::from(*secret);
    x25519_dalek::PublicKey::from(&secret).to_bytes()
}

/// The X25519 agreement. An all-zero agreement, which is what a small-order
/// point gives, is refused as `None`.
pub fn agree(secret: &[u8; 32], their_pk: &[u8; 32]) -> Option<[u8; 32]> {
    let secret = x25519_dalek::StaticSecret::from(*secret);
    let shared = secret.diffie_hellman(&x25519_dalek::PublicKey::from(*their_pk));
    shared.was_contributory().then(|| shared.to_bytes())
}

/// The message cipher's key and nonce: forty-four bytes of HKDF-SHA-256 over
/// the agreement as it stands, empty salt, `quo-seal`, key first.
pub fn seal_key(shared: &[u8; 32]) -> ([u8; 32], [u8; 12]) {
    let mut out = [0u8; 44];
    hkdf_sha256(&[], shared, LABEL_SEAL, &mut out);
    let mut key = [0u8; 32];
    let mut nonce = [0u8; 12];
    key.copy_from_slice(&out[..32]);
    nonce.copy_from_slice(&out[32..]);
    (key, nonce)
}

/// The body cipher's key and nonce: forty-four bytes of HKDF-SHA-256 over the
/// agreement then the edge key, sixty-four bytes, empty salt, `quo-edge-seal`.
pub fn edge_seal_key(shared: &[u8; 32], edge: &[u8; 32]) -> ([u8; 32], [u8; 12]) {
    let mut ikm = [0u8; 64];
    ikm[..32].copy_from_slice(shared);
    ikm[32..].copy_from_slice(edge);
    let mut out = [0u8; 44];
    hkdf_sha256(&[], &ikm, LABEL_EDGE_SEAL, &mut out);
    let mut key = [0u8; 32];
    let mut nonce = [0u8; 12];
    key.copy_from_slice(&out[..32]);
    nonce.copy_from_slice(&out[32..]);
    (key, nonce)
}

/// The next edge key: thirty-two bytes of HKDF-SHA-256 over the edge key the
/// ask came under then the reply's agreement, sixty-four bytes, empty salt,
/// `quo-edge`.
pub fn next_edge(edge: &[u8; 32], shared: &[u8; 32]) -> [u8; 32] {
    let mut ikm = [0u8; 64];
    ikm[..32].copy_from_slice(edge);
    ikm[32..].copy_from_slice(shared);
    let mut out = [0u8; 32];
    hkdf_sha256(&[], &ikm, LABEL_EDGE, &mut out);
    out
}

/// The knock's edge key: thirty-two bytes of HKDF-SHA-256 over the ML-KEM
/// shared secret, empty salt, `quo-lock`.
pub fn lock_edge(shared: &[u8; 32]) -> [u8; 32] {
    let mut out = [0u8; 32];
    hkdf_sha256(&[], shared, LABEL_LOCK, &mut out);
    out
}

/// The encapsulation key of the lock drawn as `d || z`:
/// `ML-KEM.KeyGen_internal(d, z)`.
pub fn lock_pk(seed: &[u8; LOCK_SEED_LEN]) -> Box<[u8; LOCK_LEN]> {
    let (_, ek) = lock_pair(seed);
    let mut out = Box::new([0u8; LOCK_LEN]);
    out.copy_from_slice(ek.as_bytes().as_slice());
    out
}

fn lock_pair(seed: &[u8; LOCK_SEED_LEN]) -> (<MlKem768 as KemCore>::DecapsulationKey, <MlKem768 as KemCore>::EncapsulationKey) {
    let d = B32::try_from(&seed[..32]).expect("thirty-two bytes");
    let z = B32::try_from(&seed[32..]).expect("thirty-two bytes");
    MlKem768::generate_deterministic(&d, &z)
}

/// `ML-KEM.Encaps_internal(lock, m)`: the ciphertext and the shared secret.
pub fn encapsulate(lock: &[u8; LOCK_LEN], m: &[u8; 32]) -> (Box<[u8; CIPHERTEXT_LEN]>, [u8; 32]) {
    let encoded = Encoded::<<MlKem768 as KemCore>::EncapsulationKey>::try_from(&lock[..]).expect("an encapsulation key's length");
    let ek = <MlKem768 as KemCore>::EncapsulationKey::from_bytes(&encoded);
    let (ciphertext, shared) = ek.encapsulate_deterministic(&B32::from(*m)).expect("encapsulation does not fail");
    let mut out = Box::new([0u8; CIPHERTEXT_LEN]);
    out.copy_from_slice(ciphertext.as_slice());
    (out, shared.as_slice().try_into().expect("thirty-two bytes"))
}

/// `ML-KEM.Decaps(dk, c)` under the lock drawn as `d || z`, with FIPS 203's
/// implicit rejection: a ciphertext that is not the sender's yields another
/// secret.
pub fn decapsulate(seed: &[u8; LOCK_SEED_LEN], ciphertext: &[u8; CIPHERTEXT_LEN]) -> [u8; 32] {
    let (dk, _) = lock_pair(seed);
    let c = Ciphertext::<MlKem768>::try_from(&ciphertext[..]).expect("a ciphertext's length");
    let shared = dk.decapsulate(&c).expect("decapsulation does not fail");
    shared.as_slice().try_into().expect("thirty-two bytes")
}

/// AES-256-GCM under the key and nonce drawn from `shared`, `additional` as
/// the authenticated data, the tag at the end.
pub fn encrypt(shared: &[u8; 32], additional: &[u8], plaintext: &[u8]) -> Vec<u8> {
    encrypt_under(seal_key(shared), additional, plaintext)
}

/// AES-256-GCM under a key and nonce already derived.
pub fn encrypt_under((key, nonce): ([u8; 32], [u8; 12]), additional: &[u8], plaintext: &[u8]) -> Vec<u8> {
    let mut buffer = plaintext.to_vec();
    Aes256Gcm::new(&key.into()).encrypt_in_place(Nonce::<Aes256Gcm>::from_slice(&nonce), additional, &mut buffer).expect("AES-GCM encrypts any message under the size");
    buffer
}

/// Opens what [`encrypt`] made. `None` is a box that does not open.
pub fn decrypt(shared: &[u8; 32], additional: &[u8], ciphertext: &[u8]) -> Option<Vec<u8>> {
    decrypt_under(seal_key(shared), additional, ciphertext)
}

/// Opens what [`encrypt_under`] made. `None` is a box that does not open.
pub fn decrypt_under((key, nonce): ([u8; 32], [u8; 12]), additional: &[u8], ciphertext: &[u8]) -> Option<Vec<u8>> {
    let mut buffer = ciphertext.to_vec();
    Aes256Gcm::new(&key.into()).decrypt_in_place(Nonce::<Aes256Gcm>::from_slice(&nonce), additional, &mut buffer).ok()?;
    Some(buffer)
}
